### PR TITLE
Context interp dev

### PR DESCRIPTION
**Pull request by boxofpasta**

### CODE REVIEW

File: context_interp/vgg19_features/vgg19_features.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import tensorflow as tf
+from context_interp.vgg19_features.model.model import Vgg19
+
+
+class Vgg19Features:
+
```
Comment (NeedsMorePie) [inline context_interp/vgg19_features/vgg19_features.py]: There's a blank line here.

File: context_interp/vgg19_features/vgg19_features.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    def load_pretrained_weights(self):
+        vgg19_data = Vgg19.load_params_dict(load_small=True)
+        self.vgg19 = Vgg19(data_dict=vgg19_data)
```
Comment (NeedsMorePie) [inline context_interp/vgg19_features/vgg19_features.py]: We should define member variables in a constructor and at least set it to None.

File: context_interp/vgg19_features/vgg19_features.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    def get_context_features(self, images):
+        """
+        :param images: A Tensor. Of shape [batch, H, W, num_features].
+        :return: A Tensor. Of shape [batch, H, W, 64].
+        """
+        image_features, _ = self.vgg19.build_up_to_conv1_2(images, trainable=False)
```
Comment (NeedsMorePie) [inline context_interp/vgg19_features/vgg19_features.py]: Huh, this only uses one feature map? Why not have the option to use more? Same question for the perceptual features.

File: data/interp/davis/davis_data.py [hunk 4]
```diff
@@ -2,23 +2,38 @@
...
-        with open(filename, 'rb') as fp:
-            return fp.read()
+
+        # https://stackoverflow.com/questions/31826335/how-to-convert-pil-image-image-object-to-base64-string
+        buffered = BytesIO()
+        im = Image.open(filename)
+        width, height = im.size
+
+        # This just takes the whole image currently.
+        crop_width, crop_height = width, height
```
Comment (NeedsMorePie) [inline data/interp/davis/davis_data.py]: This discards the rest of the image? Should it also take crops of the 4 corners if the crop size is small enough?

File: data/interp/interp_data_reader.py [hunk 5]
```diff
@@ -117,17 +125,16 @@ def _parse_function(example_proto):
             shot_bytes = tf.sparse_tensor_to_dense(parsed_features[SHOT], default_value=tf.as_string(0))
             shot = tf.map_fn(lambda bytes: tf.image.decode_image(bytes), shot_bytes, dtype=(tf.uint8))
             shot = tf.image.convert_image_dtype(shot, tf.float32)
-            shot = tf.reshape(shot, [shot_len, H, W, 3])
+            shot = tf.reshape(shot, (shot_len, H, W, 3))
 
+            # TODO We should randomly augment triplets with timestamps like [0.0, 1.0, 1.0], or [0.0, 0.0, 1.0].
```
Comment (NeedsMorePie) [inline data/interp/interp_data_reader.py]: Make it "TODO:".

File: mains/train_context_interp.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import argparse
+import os
+import tensorflow as tf
+from data.interp.davis.davis_data import DavisDataSet
+from context_interp.model import ContextInterp
+from train.context_interp.trainer import ContextInterpTrainer
+
+
+def add_args(parser):
```
Comment (NeedsMorePie) [inline mains/train_context_interp.py]: Let's put this below the main() function for consistency.

File: mains/train_context_interp.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        os.makedirs(args.checkpoint_directory)
+
+    print('Creating network...')
+    model = ContextInterp()
+
+    print('Creating dataset...')
+    dataset = DavisDataSet(args.directory, [[1]],
+                           batch_size=args.batch_size)
+
+    # TODO Some of this stuff might want to go into a config json.
```
Comment (NeedsMorePie) [inline mains/train_context_interp.py]: "TODO:"

File: train/context_interp/trainer.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        # Get the optimizer.
+        with tf.variable_scope('train'):
+            self.global_step = tf.Variable(initial_value=0, trainable=False, dtype=tf.int32, name='global_step')
+            self.train_op = AdamaxOptimizer(config['learning_rate'], beta1=0.9, beta2=0.999).minimize(
+                self.loss, global_step=self.global_step
+            )
+
+        # Checkpoint saving.
+        self.saver = tf.train.Saver()
+        self.valid_logger = Logger(os.path.join(self.config['checkpoint_directory'], 'valid'), self.session.graph)
```
Comment (NeedsMorePie) [inline train/context_interp/trainer.py]: I am confused. This Logger util looks like it would be used for creating summaries for non-tensorflow frameworks. Why use this if we are using tensorflow? It looks like things are being reevaluated just to create np versions to be converted to tf summaries.

File: utils/tf.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    slices = tf.cond(
+        num_offsets > 0,
+        true_fn=lambda: get_slice(slice_indices_tensor, num_offsets, x),
+        false_fn=lambda: get_zeros(sequence_len, x)
+    )
+    return slices
+
+
+# Copied from: https://github.com/tensorflow/tensorflow/issues/312
+def optimistic_restore(session, save_file):
```
Comment (NeedsMorePie) [inline utils/tf.py]: What do we need this for? Is there something wrong with restoring the whole checkpoint as is?

File: README.md [hunk 10]
```diff
@@ -9,6 +9,7 @@
     *   opencv-python
     *   matplotlib
     *   joblib
+    *   pillow
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply README.md]: In imports it will show up as PIL. It's very commonly used for image processing.

File: context_interp/gridnet/connections/connections.py [hunk 11]
```diff
@@ -35,12 +36,17 @@ def get_forward(self, features, reuse_variables=False, training=False):
...
-            # Pass through resolution preserving convolutions, with skip connection at the end.
+            # Pass through resolution preserving convolutions.
             if self.activation_fn is not None:
                 previous_output = self.activation_fn(features)
 
             previous_output, layer_outputs = self._get_conv_tower(previous_output)
-            final_output = features + previous_output
+
+            # Add skip connection only if channels are the same size.
+            if features.get_shape().as_list()[-1] == previous_output.get_shape().as_list()[-1]:
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/gridnet/connections/connections.py]: This one requires specifically the number of channels to be statically known, which it should be. I think it's ok for the other members of the shape list to be None.

File: context_interp/model.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                 flow_b_a: Flow from b to a (centered at b).
+        """
+        with tf.variable_scope(self.name, reuse=reuse_variables):
+            image_a_contexts = self.feature_extractor.get_context_features(image_a)
+            image_b_contexts = self.feature_extractor.get_context_features(image_b)
+
+            # TODO Add instance normalization. Described in 3.3 of https://arxiv.org/pdf/1803.10967.pdf.
+
+            # Get a->b and b->a flows from PWCNet.
+            flow_a_b, _ = self.pwcnet.get_forward(image_a, image_b)
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/model.py]: I've thought of doing so, and I'm not entirely convinced of performance benefits with this change after our experiments with the cost volume; we might have enough large tensor memory issues as is. Perhaps I should setup another forward + backward pass profiling test. Aside from performance, having them separated out is preferred as the code and the graphs are more readable.

File: context_interp/model.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            warped_combined = tf.stop_gradient(warped_combined)
+            synthesized, _, _, _ = self.gridnet.get_forward(warped_combined, training=True)
+            return synthesized, warped_a_b, warped_b_a, flow_a_b, flow_b_a
+
+    def load_pwcnet_weights(self, pwcnet_weights_path, sess):
+        """
+        Loads pre-trained PWCNet weights.
+        For this to work:
+            - It must be called after get_forward.
+            - get_forward must not have been called in an enclosing variable scope.
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/model.py]: It's still true, since scope_prefix only uses self.name. An enclosing scope during variable definitions would not work well with this.

File: context_interp/model_test.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        interpolated_tensor = output_tensors[0]
+
+        image_a = np.zeros(shape=[batch_size, height, width, im_channels], dtype=np.float32)
+        image_b = np.zeros(shape=[batch_size, height, width, im_channels], dtype=np.float32)
+        image_a[:, 2:height-2, 2:width-2, :] = 1.0
+        image_b[:, 4:height-4, 5:width-5, :] = 1.0
+
+        loss_tensor = model.get_training_loss(interpolated_tensor, gt_placeholder)
+
+        # Currently a stop_gradient is applied to the input of GridNet.
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/model_test.py]: We might want to re-enable this if pre-warping is allowed to have gradients flowing through.

File: context_interp/model_test.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        # Check shapes.
+        self.assertTrue(np.allclose(interpolated.shape, np.asarray([batch_size, height, width, im_channels])))
+        self.assertTrue(np.allclose(warped_im_a_b.shape, np.asarray([batch_size, height, width, im_channels])))
+        self.assertTrue(np.allclose(warped_im_b_a.shape, np.asarray([batch_size, height, width, im_channels])))
+        self.assertTrue(np.allclose(warped_feat_a_b.shape[:-1], np.asarray([batch_size, height, width])))
+        self.assertTrue(np.allclose(warped_feat_b_a.shape[:-1], np.asarray([batch_size, height, width])))
+        self.assertTrue(np.allclose(warped_feat_a_b.shape, warped_feat_b_a.shape))
+
+        # Check the gradients with the loss.
+        grad_tensors = tf.gradients(loss_tensor, interpolated_tensor)
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/model_test.py]: We could, it's just that that variable wasn't exposed in the return of get_forward (and is pretty much useless otherwise). Besides, as the pre-warping stuff has no gradients flowing through it, the GridNet tests should cover the rest.

File: context_interp/vgg19_features/vgg19_features.py (hunk 3 above)
Comment (boxofpasta) [reply context_interp/vgg19_features/vgg19_features.py]: Do you want it to return all the layers of VGG that are of the given resolution? The only reason for not doing more is simply that it wasn't being done in the paper, and I wasn't sure what we'd want to do more of.

File: data/interp/davis/davis_data.py (hunk 4 above)
Comment (boxofpasta) [reply data/interp/davis/davis_data.py]: This piece of cropping code here doesn't do anything. I just kept it for future reference, in case I do want to crop or resize images before saving to TFRecords.

File: mains/create_davis_dataset.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+            <video_shot_2>
+            ...
+            <video_shot_n>
+    """
+    parser = argparse.ArgumentParser()
+    add_args(parser)
+    args = parser.parse_args()
+
+    input_directory = os.path.join(args.directory, 'JPEGImages', '480p')
+    tf_records_directory = os.path.join(args.out_directory)
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply mains/create_davis_dataset.py]: It's there to make people laugh.

File: mains/create_davis_dataset.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+    dataset = DavisDataSet(tf_records_directory, [[1]])
+    dataset.set_verbose(True)
+    dataset.preprocess_raw(input_directory, args.shard_size, validation_size=args.num_validation)
+
+
+def add_args(parser):
+    parser.add_argument('-d', '--directory', type=str,
+                        help='Directory of the unzipped DAVIS dataset.')
+    parser.add_argument('-o', '--out_directory', type=str,
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply mains/create_davis_dataset.py]: I think it's cleaner to specify an empty / non-existent folder that will be used just to keep the TFRecords. If you want to mix everything you should be able to, but I don't like that as a default.

File: utils/tf.py (hunk 9 above)
Comment (NeedsMorePie) [reply utils/tf.py]: This needs more comments.

File: context_interp/model_test.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (NeedsMorePie) [reply context_interp/model_test.py]: If this much code is important, then it can't just sit here as dead code. Wrap it in an if statement or something.

File: context_interp/model_test.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (NeedsMorePie) [reply context_interp/model_test.py]: Don't you have the warped images in the tensorboard visualizations? We can test that the gradients flow to them.

File: context_interp/vgg19_features/vgg19_features.py (hunk 3 above)
Comment (NeedsMorePie) [reply context_interp/vgg19_features/vgg19_features.py]: Experiments on that should be done later so maybe add a TODO on this.

File: data/interp/davis/davis_data.py (hunk 4 above)
Comment (NeedsMorePie) [reply data/interp/davis/davis_data.py]: But if it were to do something, I think leaving out significant portions of the data would be sort of a data bug. I think it should either be fixed or be treated as dead code and deleted.

File: mains/create_davis_dataset.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (NeedsMorePie) [reply mains/create_davis_dataset.py]: That's fine but it is now inconsistent with the flow dataset's commands. I am bothered more by inconsistency (even if its temporary) than defaulting to saving in the same directory.

File: context_interp/model.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (NeedsMorePie) [reply context_interp/model.py]: The reason the cost volume was slower is probably unrelated. I have definitely dealt with reusing variables with new ops in networks before and there is a performance difference -- this is not speculation. The tensorboard cleanliness definitely doesn't come before actual performance gains.

File: context_interp/model.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (NeedsMorePie) [reply context_interp/model.py]: The fact that scope_prefix only uses self.name is a self-fulfilling prophecy -- that's just how you wrote this function. Wouldn't it be more useful to add a comment on how changes to the topography affect this function?

File: train/context_interp/trainer.py (hunk 8 above)
Comment (NeedsMorePie) [reply train/context_interp/trainer.py]: That's fine I guess. Though it does appear to be quite a bit of code to solve something that isn't much of a problem anyway.

File: README.md [hunk 18]
```diff
@@ -121,3 +122,59 @@ python -m unittest pwcnet.warp.warp_test
...
+    ```
+    
+#### Training
+
+1.  Have your tf records prepared.
+
+2.  Run the following command:
+
+    ```
+    python -m mains.train_context_interp -d path/to/tfrecords_dir -c path/to/checkpoints_dir -w path/to/pwcnet_weights.npz
```
[the remark this answers is not in the document]
Comment (NeedsMorePie) [reply README.md]: By that I mean mention in step 1 that you need a trained pwcnet to get any further.

File: train/context_interp/trainer.py (hunk 8 above)
Comment (boxofpasta) [reply train/context_interp/trainer.py]: The validation summaries weren't working properly (not displaying enough variety of images, way too many unnecessary saves, loss plot was messed up) so I think it's a necessary fix.

File: context_interp/vgg19_features/vgg19_features.py (hunk 3 above)
Comment (boxofpasta) [reply context_interp/vgg19_features/vgg19_features.py]: This is not something that's as concrete as other TODO tasks, it's more like an optional thing to try -- we will have many more of these things soon, and I want to keep tabs on these elsewhere (perhaps in the wiki?).

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (boxofpasta) [reply README.md]: I forgot to update this last night.

File: mains/create_davis_dataset.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (boxofpasta) [reply mains/create_davis_dataset.py]: I do not want to accidentally spill TFRecords everywhere. If you are bothered by the inconsistency, it might be better for you to just change the flow dataset's command?

File: context_interp/model.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+            features_a = tf.concat([image_a, image_a_contexts], axis=-1)
+            features_b = tf.concat([image_b, image_b_contexts], axis=-1)
+
+            # Warp images and their contexts from a->b and from b->a.
+            warped_a_b = forward_warp(features_a, t * flow_a_b)
+            warped_b_a = forward_warp(features_b, (1.0 - t) * flow_b_a)
+
+            # Feed into GridNet for final synthesis.
+            warped_combined = tf.concat([warped_a_b, warped_b_a], axis=-1)
```
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/model.py]: It's in a different axis.

File: context_interp/model.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (boxofpasta) [reply context_interp/model.py]: I wrote a function and stated the pre-reqs for using it, I don't see anything wrong with that in itself. I have updated this function so that it doesn't have that particular requirement though.
